File: cs50/sentimental-readability/readability.py
from cs50 import get_string


def calculateLS(text):
    letters = 0
    words = 0
    sentences = 0
    in_word = False

    for char in text:
        if char.isalpha():
            letters += 1
        if char.isspace():
            if in_word:
                words += 1
                in_word = False
        else:
            in_word = True
        if char in ['.', '!', '?']:
            sentences += 1

    if in_word:
        words += 1

    if words > 0:
        L = (letters / words) * 100
        S = (sentences / words) * 100
    else:
        L = 0
        S = 0

    return L, S


text = get_string("Text: ")
L, S = calculateLS(text)
index = 0.0588 * L - 0.296 * S - 15.8
rounded_index = round(index)

if rounded_index >= 16:
    print("Grade 16+")
elif rounded_index < 1:
    print("Before Grade 1")
else:
    print(f"Grade {rounded_index}")

File: cs50/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

void calculateLS(const char *text, double *L, double *S);

int main(void)
{
    char *text = get_string("Text: ");
    double L, S;
    calculateLS(text, &L, &S);
    double index = 0.0588 * L - 0.296 * S - 15.8;
    int rounded_index = round(index);
    if (rounded_index >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (rounded_index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %d\n", rounded_index);
    }
    return 0;
}

void calculateLS(const char *text, double *L, double *S)
{
    int letters = 0, words = 0, sentences = 0, i = 0, inWord = 0;
    while (text[i] != '\0')
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
        if (isspace(text[i]))
        {
            if (inWord)
            {
                words++;
                inWord = 0;
            }
        }
        else
        {
            inWord = 1;
        }
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentences++;
        }
        i++;
    }
    if (inWord)
    {
        words++;
    }
    if (words > 0)
    {
        *L = (letters / (double) words) * 100;
        *S = (sentences / (double) words) * 100;
    }
    else
    {
        *L = 0;
        *S = 0;
    }
}
